draft-trammell-taps-post-sockets.md

```diff
@@ -318,40 +318,40 @@ interaction patterns, defined in the subsections below.
-
-   Writing a byte to a Stream will cause it to be received by the remote, in
-   order, or will cause an error condition and termination of the stream if the
-   byte cannot be delivered. Due to the strong sequential dependence on a stream,
-   streams must always be reliable and ordered. A Message Carrier may only be
-   morphed to a Stream if it uses transport protocol stack that provides
-   reliable, ordered service, and only before it is used to send a Message.
-
 ## Message
 
-A Message is an atomic unit of communication between applications. A Message
-that cannot be delivered in its entirety within the constraints of the network
-connectivity and the requirements of the application is not delivered at all.
-
-Messages can represent both relatively small structures, such as requests in a
+A Message is the unit of communication between applications. Messages
+can represent both relatively small structures, such as requests in a
 request/response protocol such as HTTP; as well as relatively large
 structures, such as files of arbitrary size in a filesystem.
 
 In the general case, there is no mapping between a Message and packets sent by
 the underlying protocol stack on the wire: the transport protocol may freely
 segment messages and/or combine messages into packets. However, a message may be
-marked as immediate, which will cause it to be sent in a single packet, if it
-will fit.
-
-This implies that both the sending and receiving endpoint, whether in the
-application layer or the transport layer, must guarantee storage for the full
-size of an Message.
+marked as immediate, which will cause it to be sent in a single packet when
+possible.
+
+Messages may be sent and received either atomically, or non-atomically. All
+protocols must support non-atomic Messages, but only protocols that have
+fixed maximum lengths may allow clients to treat Messages as atomic.
+
+When a Message is received atomically, the entire content of the Message must
+be delivered at once, and the Message is not delivered at all if the full Message
+is not received.  This implies that both the sending and receiving endpoint,
+whether in the application or the carrier, must guarantee storage for the full
+size of a Message. Dealing with atomic Messages should be preferred for
+simplicity whenever possible based on the underlying protocol.
+
+When Messages are handled non-atomically, they may be sent or received in
+several stages, with a handle representing the total Message being associated
+with each portion of the content. Each call to send or receive also indicates
+whether or not the Message is now complete. This approach is necessary whenever
+the size of the message does not have a known bound, or the size is too large to
+process and hold in memory atomically. Protocols that only present a concept
+of byte streams represent their data as single Messages with unknown bounds.
+In the case of TCP, the client will receive a single Message in non-atomic pieces,
```

Comment:
"In the case of TCP" -> "In the case of a raw TCP stream". The API could expose HTTP over TCP as a protocol, and that protocol could do the fragmentation and assembly and expose atomic messages.